src/function/matrix/eigs.js

```diff
@@ -41,63 +54,69 @@ export const createEigs = /* #__PURE__ */ factory(name, dependencies, ({ config,
    * @param {Array | Matrix} x  Matrix to be diagonalized
    *
    * @param {number | BigNumber} [prec] Precision, default value: 1e-15
-   * @return {{values: Array|Matrix, vectors: Array|Matrix}} Object containing an array of eigenvalues and a matrix with eigenvectors as columns.
+   * @return {{values: Array|Matrix, eigenvectors: Array<EVobj>}} Object containing an array of eigenvalues and an array of {value: number|BigNumber, vector: Array|Matrix} objects.
    *
    */
   return typed('eigs', {
 
-    Array: function (x) {
-      const mat = matrix(x)
-      return computeValuesAndVectors(mat)
-    },
-
+    // The conversion to matrix in the first two implementations,
+    // just to convert back to an array right away in
+    // computeValuesAndVectors, is unfortunate, and should perhaps be
+    // streamlined. It is done because the Matrix object carries some
+    // type information about its entries, and so constructing the matrix
+    // is a roundabout way of doing type detection.
+    Array: function (x) { return computeValuesAndVectors(matrix(x)) },
     'Array, number|BigNumber': function (x, prec) {
-      const mat = matrix(x)
-      return computeValuesAndVectors(mat, prec)
+      return computeValuesAndVectors(matrix(x), prec)
     },
-
     Matrix: function (mat) {
-      const { values, vectors } = computeValuesAndVectors(mat)
-      return {
-        values: matrix(values),
-        vectors: matrix(vectors)
-      }
+      return computeValuesAndVectors(mat, undefined, true)
     },
-
     'Matrix, number|BigNumber': function (mat, prec) {
-      const { values, vectors } = computeValuesAndVectors(mat, prec)
-      return {
-        values: matrix(values),
-        vectors: matrix(vectors)
-      }
+      return computeValuesAndVectors(mat, prec, true)
     }
   })
 
-  function computeValuesAndVectors (mat, prec) {
+  function computeValuesAndVectors (mat, prec, matricize = false) {
     if (prec === undefined) {
       prec = config.epsilon
     }
 
-    const size = mat.size()
+    let result
+    const arr = mat.toArray()
+    const asize = mat.size()
 
-    if (size.length !== 2 || size[0] !== size[1]) {
-      throw new RangeError('Matrix must be square (size: ' + format(size) + ')')
+    if (asize.length !== 2 || asize[0] !== asize[1]) {
+      throw new RangeError(`Matrix must be square (size: ${format(asize)})`)
     }
 
-    const arr = mat.toArray()
-    const N = size[0]
+    const N = asize[0]
 
     if (isReal(arr, N, prec)) {
       coerceReal(arr, N)
```

Comment:
I know this code was already there, so no regression, but I think `coerceReal(arr, N)` can mutate the original input data. I think we should clone the data first before mutating it.

Reply:
arr is the "working memory" for the algorithm, intended to be modified in place for efficiency. It was produced by mat.toArray(), so is already guaranteed to be unaliased and safe to modify in place. I will put a comment to that effect.

Reply:
I was assuming that `toArray` wouldn't copy the contents, but it actually does, so there is indeed no problem 😅 .

https://github.com/josdejong/mathjs/blob/develop/src/type/matrix/DenseMatrix.js#L665-L667